src/ctap/storage/key.rs

```diff
@@ -0,0 +1,135 @@
+#[cfg(test)]
+macro_rules! make_range {
+    ($key: literal..$end: literal) => {
+        $key..$end
+    };
+    ($key: literal) => {
+        $key..$key + 1
+    };
+}
+
+/// Helper to define keys as a partial partition of a range.
+macro_rules! make_partition {
+        ($range: expr,
+         $(
+             $(#[$doc: meta])*
+             $name: ident = $key: literal $(.. $end: literal)?;
+         )*) => {
+            $(
+                make_key!($(#[$doc])* $name = $key $(.. $end)?);
+            )*
+            #[cfg(test)]
+            const KEY_RANGE: core::ops::Range<usize> = $range;
+            #[cfg(test)]
+            const ALL_KEYS: &[core::ops::Range<usize>] = &[$(make_range!($key $(.. $end)?)),*];
+        };
+    }
+
+make_partition! {
+    // We reserve 0 and 2048+ for possible migration purposes. We add persistent entries starting
+    // from 1 and going up. We add non-persistent entries starting from 2047 and going down. This
+    // way, we don't commit to a fixed number of persistent keys.
+    1..2048,
+
+    // WARNING: Keys should not be deleted but prefixed with `_` to avoid accidentally reusing them.
+
+    /// The attestation private key.
+    ATTESTATION_PRIVATE_KEY = 1;
+
+    /// The attestation certificate.
+    ATTESTATION_CERTIFICATE = 2;
+
+    /// The aaguid.
+    AAGUID = 3;
+
+    // This is the persistent key limit:
+    // - When adding a (persistent) key above this message, make sure its value is smaller than
+    //   NUM_PERSISTENT_KEYS.
+    // - When adding a (non-persistent) key below this message, make sure its value is bigger or
+    //   equal than NUM_PERSISTENT_KEYS.
+
+    /// The credentials.
+    ///
+    /// Depending on `MAX_SUPPORTED_RESIDENTIAL_KEYS`, only a prefix of those keys is used. Each
+    /// board may configure `MAX_SUPPORTED_RESIDENTIAL_KEYS` depending on the storage size.
+    CREDENTIALS = 1700..2000;
+
+    /// TODO: Help from reviewer needed for documentation.
+    _MIN_PIN_LENGTH_RP_IDS = 2042;
+
+    /// The minimum PIN length.
+    #[cfg(feature = "with_ctap2_1")]
+    MIN_PIN_LENGTH = 2043;
+
+    /// The number of PIN retries.
+    ///
+    /// If the entry is absent, the number of PIN retries is `MAX_PIN_RETRIES`.
+    PIN_RETRIES = 2044;
+
+    /// The PIN hash.
+    ///
+    /// If the entry is absent, there is no PIN set.
+    PIN_HASH = 2045;
+
+    /// The encryption and hmac keys.
+    ///
+    /// This entry is always present. It is generated at startup if absent. This is not a persistent
+    /// key because its value should change after a CTAP reset.
+    MASTER_KEYS = 2046;
+
+    /// The global signature counter.
+    ///
+    /// If the entry is absent, the counter is 0.
+    GLOBAL_SIGNATURE_COUNTER = 2047;
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn enough_credentials() {
+        use super::super::MAX_SUPPORTED_RESIDENTIAL_KEYS;
+        assert!(MAX_SUPPORTED_RESIDENTIAL_KEYS <= CREDENTIALS.end - CREDENTIALS.start);
```

Comment:
One advantage I see of keeping this constant is to decorrelate store format and design requirements. In particular, having more "room" for credential keys in the store permits to increase the number of supported credentials later without having to make the set of credential keys discontinuous.

The things to have in mind are:
- Limiting the number of credentials is useful for lifetime planning. It permits to guarantee a minimum amount of increments. This is what I call "design requirement".
- The key partition defined in this file cannot be easily revised without a migration. The only supported operation is to define a new key and deprecating an existing key. Changing the meaning of an existing or deprecated key is not compatible (unless we can ensure that key was never used) and requires an explicit migration process.

Reply:
My concern about migration is firmware update. We might want in a few years to say that users can store more credentials. We would release a new binary with a new constant. If this new constant is below what the current key partition supports, there is nothing else to do. However if the constant is bigger then we need to either:
- Create an additional range for credentials (probably the best option, it would just complicate a bit the code to look into both ranges).
- Migrate the store to use a new key partition (i.e. some defragmentation of the solution above).

Regarding the dynamic guessing, we can indeed do that (you can create an issue and I'll do it). However, we would still need some compile-time constant to say how much capacity we want to allocate to credentials (such that we have enough remaining for counters and increments). We would then have 2 limits for credentials:
1. How many keys we allocated (currently 300 but we can change it in this PR without compatibility issues).
2. How much capacity credentials can hold in the store (some compile-time constant that can be increased in the future without store migration).

Reply:
No need to change the logic here, I'll wait for 2.1 to settle.

Concerning the number of credentials, not sure what the best value is. The average user has 100 passwords in 2020, I don't know about the tail of this distribution. Other security keys usually stored 25-100 credentials.

I don't expect passwordless and resident keys to be big in the next few years, but storing credentials should still be the main task of this storage. So if we have 4k keys available, I don't see why we wouldn't go for more, say 500. But that's just guessing...

Reply:
![why_not_do_both](https://memegenerator.net/img/instances/38166116/why-not-do-both.jpg)

We can actually do the following:
- We use 300 (or whatever number C1) keys for credentials.
- But we also reserve contiguously 200 (or whatever number C2) keys for either more credentials or other things.

This means we essentially have C1 + C2 keys for credentials if needed but we don't commit for C2 of them.

Concretely, We currently have credentials use 1700..2000, we could add 1500..1700 as a reserved range (i.e. not used but allocated). New keys would be added 1499 and below.

This way, in the future, we can extend the number of credentials with a simple firmware update. And it also doesn't block us if we need this range for other things. The things I have in mind that could consume many keys are credential-associated data with different read-write pattern like per-credential counters (those are written as often as read, while credentials are much more read than written) to save flash lifetime. We should have a copy of the credential range for those counters, so we can easily double our used keys (and even more if we have other use-cases). Also note that from the 4k keys, I already reserve 2k for migration purposes. But I think 500 keys for credentials sound alright. So we could do any of:
1. Use 300 keys but reserve 200 contiguously in case. (I slightly prefer this because it postpones the decision to when we have better data.)
2. Use 500 keys upfront.

Actually, we could even reserve 1000..1700 (700 keys) such that 1000..2000 is reserved for credential-related stuff. Credentials consume from 2000 downward. Possible per-credential counters if they exist one day would consume from 1000 upward. If they don't then this space could be used for other things.

Reply:
Reserving 1000..2000 for credential related keys sounds like the perfect solution.

Reply:
Done.